File: src/main/java/com/letgo/book/infrastructure/persistence/mapping/BookIdDescriptor.java
package com.letgo.book.infrastructure.persistence.mapping;

import com.letgo.book.domain.BookId;
import org.hibernate.type.descriptor.WrapperOptions;
import org.hibernate.type.descriptor.java.AbstractTypeDescriptor;
import org.hibernate.type.descriptor.java.ImmutableMutabilityPlan;

final public class BookIdDescriptor extends AbstractTypeDescriptor<BookId> {
    public BookIdDescriptor() {
        super(BookId.class, new ImmutableMutabilityPlan<>());
    }

    @Override
    public String toString(BookId value) {
        return value.value();
    }

    @Override
    public BookId fromString(String string) {
        return BookId.create(string);
    }

    @Override
    public <X> X unwrap(BookId value, Class<X> type, WrapperOptions options) {
        if (value == null) {
            return null;
        }

        if (String.class.isAssignableFrom(type)) {
            return (X) toString(value);
        }

        throw unknownUnwrap(type);
    }

    @Override
    public <X> BookId wrap(X value, WrapperOptions options) {
        if (value == null) {
            return null;
        }

        if (value instanceof String) {
            return fromString((String) value);
        }

        throw unknownWrap(value.getClass());
    }
}

File: src/main/java/com/letgo/book/infrastructure/persistence/mapping/BookIdDescriptor.kt
package com.letgo.book.infrastructure.persistence.mapping

import com.letgo.book.domain.BookId
import org.hibernate.type.descriptor.WrapperOptions
import org.hibernate.type.descriptor.java.AbstractTypeDescriptor
import org.hibernate.type.descriptor.java.ImmutableMutabilityPlan

class BookIdDescriptor : AbstractTypeDescriptor<BookId>(BookId::class.java, ImmutableMutabilityPlan()) {
    override fun toString(value: BookId): String {
        return value.value()
    }

    override fun fromString(string: String): BookId {
        return BookId.create(string)
    }

    override fun <X> unwrap(value: BookId?, type: Class<X>, options: WrapperOptions): X? {
        if (value == null) {
            return null
        }
        if (String::class.java.isAssignableFrom(type)) {
            return toString(value) as X
        }
        throw unknownUnwrap(type)
    }

    override fun <X> wrap(value: X?, options: WrapperOptions): BookId? {
        if (value == null) {
            return null
        }
        if (value is String) {
            return fromString(value as String)
        }
        throw unknownWrap(value.javaClass)
    }
}
